--- netconfig.pp.orig	2015-02-26 10:16:48.455129093 +0000
+++ netconfig.pp	2015-02-26 10:35:36.248132567 +0000
@@ -1,12 +1,12 @@
 notice('MODULAR: netconfig.pp')
 
-if  hiera('disable_offload') {
-  L23network::L3::Ifconfig<||> {
-    ethtool => {
-      'K' => ['gso off',  'gro off'],
-    }
-  }
-}
+#if  hiera('disable_offload') {
+#  L23network::L3::Ifconfig<||> {
+#    ethtool => {
+#      'K' => ['gso off',  'gro off'],
+#    }
+#  }
+#}
 
 class { 'l23network' :
   use_ovs => hiera('use_neutron'),
