--- fabfile/tasks/ha.py.orig	2015-06-25 12:59:50.256028918 +0000
+++ fabfile/tasks/ha.py	2015-06-25 13:18:00.060015103 +0000
@@ -222,8 +222,8 @@
 
 def setup_keepalived_node(role):
     """Task to provision VIP for node with keepalived"""
-    mgmt_ip = hstr_to_ip(env.host_string)
-    self_host = get_control_host_string(env.host_string)
+    mgmt_ip = env.hoststring_mgm_ip[env.host_string]
+    self_host = get_control_host_string(env.host_string)
     self_ip = hstr_to_ip(self_host)
 
     internal_vip = get_openstack_internal_vip()
