diff -u /opt/contrail/utils/fabfile/tasks/ha.py /opt/contrail/utils/fabfile/tasks/ha.py
--- /opt/contrail/utils/fabfile/tasks/ha.py
+++ /opt/contrail/utils/fabfile/tasks/ha.py
@@ -199,9 +199,12 @@
 
 def setup_keepalived_node(role):
     """Task to provision VIP for node with keepalived"""
-    mgmt_ip = hstr_to_ip(env.host_string)
     self_host = get_control_host_string(env.host_string)
-    self_ip = hstr_to_ip(self_host)
+
+    # looks like smth messed up several levels below, so vars are swapped and
+    # it's the only way which works
+    self_ip = env.hoststring_mgm_ip[env.host_string]
+    mgmt_ip = env.hoststring_public_ip[env.host_string]
 
     internal_vip = get_openstack_internal_vip()
     external_vip = get_openstack_external_vip()
@@ -284,11 +287,11 @@
                  % (space, server_index, host_ip)
         if server_index == 1:
             rabbitmq_server_lines +=\
-                '%s server rabbit%s %s:5672 weight 200 check inter 2000 rise 2 fall 3\n'\
+                '%s server rabbit%s %s:25672 weight 200 check inter 2000 rise 2 fall 3\n'\
                  % (space, server_index, host_ip)
         else:
             rabbitmq_server_lines +=\
-                '%s server rabbit%s %s:5672 weight 100 check inter 2000 rise 2 fall 3 backup\n'\
+                '%s server rabbit%s %s:25672 weight 100 check inter 2000 rise 2 fall 3 backup\n'\
                  % (space, server_index, host_ip)
         if server_index == 1:
              mysql_server_lines +=\
