--- /opt/contrail/utils/fabfile/tasks/provision.py.orig
+++ /opt/contrail/utils/fabfile/tasks/provision.py
@@ -463,7 +463,7 @@
                     cmd += ' --manage_neutron %s' % manage_neutron
             else:
                 cmd += ' --manage_neutron no'
-            internal_vip = get_contrail_internal_vip()
+            internal_vip = get_contrail_external_vip()
             if internal_vip:
                 # Highly available setup
                 cmd += ' --internal_vip %s' % (internal_vip)
